### PR TITLE
DEV-6755: added Orders tutorial tests.

### PR DESCRIPTION
**Pull request by timbarrass**
# Pull Request Checklist

- [x] Read the [contributing guidelines](https://github.com/finbourne/lusid-sdk-python-preview/blob/master/CONTRIBUTING.md)
- [x] Tests pass
- [x] Raised the PR against the `develop` branch

# Description of the PR

Adds first set of Orders tutorial integration tests.

### CODE REVIEW

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+            Assert.That(upsertResult.Values.Count(), Is.EqualTo(3));
+            Assert.That(upsertResult.Values.Single(rl => rl.Id.Id.Equals(orderId1)).LusidInstrument, Is.EqualTo(_instrumentIds.First()));
+            
+            var quantityFilter = _ordersApi.ListOrders(scope: testScope, asAt: DateTimeOffset.UtcNow, filter: "Quantity gt 100");
+
+            Assert.That(quantityFilter.Values.Count(), Is.EqualTo(2));
+            
+            var orderBookFilter = _ordersApi.ListOrders(scope: testScope, asAt: DateTimeOffset.UtcNow, filter: $"Properties[{testScope}/OrderBook] eq 'UK Test Orders 2'");
+
+            Assert.That(orderBookFilter.Values.Count(), Is.EqualTo(1));
```
Comment (MikeMcGarry) [inline sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: Should we check that the property value on this order is correct as well?

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+            
+            var instrumentFilter = _ordersApi.ListOrders(scope: testScope, asAt: DateTimeOffset.UtcNow, filter: $"InstrumentUid eq '{_instrumentIds.First()}'");
+            
+            Assert.That(instrumentFilter.Values.Count(), Is.EqualTo(2));
```
Comment (MikeMcGarry) [inline sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: Should we check the instrument id is correct?

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+            var upsertResult = _ordersApi.UpsertOrders(scope: testScope, request: request);
+
+            // The return gives us a list of orders upserted, and the lusidinstrument for each has been mapped from the
+            // instrument identifiers passed
+            Assert.That(upsertResult.Values.Count(), Is.EqualTo(3));
+            Assert.That(upsertResult.Values.Single(rl => rl.Id.Id.Equals(orderId1)).LusidInstrument, Is.EqualTo(_instrumentIds.First()));
+            
+            var quantityFilter = _ordersApi.ListOrders(scope: testScope, asAt: DateTimeOffset.UtcNow, filter: "Quantity gt 100");
+
+            Assert.That(quantityFilter.Values.Count(), Is.EqualTo(2));
```
Comment (MikeMcGarry) [inline sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: Should we check that the quantities are actually greater than 100 on these?

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+            // from the instrument identifiers passed.
+            var request = new OrderSetRequest(
+                orderRequests: new List<OrderRequest>
+                {
+                    new OrderRequest(
+                        code: orderId,
+                        quantity: 100,
+                        instrumentIdentifiers: new Dictionary<string, string>()
+                        {
+                            ["Instrument/default/LusidInstrumentId"] = _instrumentIds.First()
```
Comment (rizwansaeed) [inline sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: Presumably these have the same resolution behaviours as transactions? i.e. they don't have to be defined as unique identifiers in LUISD?

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+                        portfolio: new ResourceId(testScope, "OrdersTestPortfolio"),
+                        properties: new List<PerpetualProperty>
+                        {
+                            new PerpetualProperty($"Order/{testScope}/TIF", new PropertyValue("GTC")),
+                            new PerpetualProperty($"Order/{testScope}/OrderBook", new PropertyValue("UK Test Orders")),
+                            new PerpetualProperty($"Order/{testScope}/PortfolioManager", new PropertyValue("F Bar")),
+                            new PerpetualProperty($"Order/{testScope}/Account", new PropertyValue("J Wilson")),
+                            new PerpetualProperty($"Order/{testScope}/Strategy", new PropertyValue("RiskArb")),
+                        },
+                        lusidInstrument: ""
```
Comment (rizwansaeed) [inline sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: what is this/why is this needed?

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+        
+        [Test]
+        public void Update_Simple_Order()
+        {
+            var testScope = $"TestScope-{Guid.NewGuid().ToString()}";
+            var orderId = $"Order-{Guid.NewGuid().ToString()}";
+            
+            // We want to make a request for a single order. The internal security id will be mapped on upsert
+            // from the instrument identifiers passed. Properties
+            var request = new OrderSetRequest(
```
Comment (rizwansaeed) [inline sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: I'd rename to `initialOrderRequest` to make it clear

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+                            new PerpetualProperty($"Order/{testScope}/Strategy", new PropertyValue("RiskArb")),
+                        },
+                        lusidInstrument: "" 
+                    )
+                });
+
+            // We can ask the Orders API to upsert this order for us
+            var upsertResult = _ordersApi.UpsertOrders(scope: testScope, request: request);
+
+            // The return gives us a list of orders upserted, and the lusidinstrument for each has been mapped from the
```
Comment (rizwansaeed) [inline sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: Does this return an Instrument or LusidInstrumentId?

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs [hunk 8]
```diff
@@ -0,0 +1,270 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Lusid.Sdk.Api;
+using Lusid.Sdk.Model;
+using Lusid.Sdk.Tests.Utilities;
+using Lusid.Sdk.Utilities;
+using NUnit.Framework;
+
+namespace Lusid.Sdk.Tests.tutorials.Ibor
```
[the remark this answers is not in the document]
Comment (timbarrass) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: Ah, it took it's capitalisation from the folder names -- done.

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+                orderRequests: new List<OrderRequest>
+                {
+                    new OrderRequest(
+                        code: orderId,
+                        quantity: 100,
+                        instrumentIdentifiers: new Dictionary<string, string>()
+                        {
+                            ["Instrument/default/LusidInstrumentId"] = _instrumentIds.First()
+                        },
+                        portfolio: new ResourceId(testScope, "OrdersTestPortfolio"),
```
[the remark this answers is not in the document]
Comment (timbarrass) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: At the moment it's not validated as existing; we need to discuss whether an Order must or can reference a Portfolio this in the context of the data model generally.

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+                        properties: new List<PerpetualProperty>
+                        {
+                            new PerpetualProperty($"Order/{testScope}/TIF", new PropertyValue("GTC")),
+                            new PerpetualProperty($"Order/{testScope}/OrderBook", new PropertyValue("UK Test Orders")),
+                            new PerpetualProperty($"Order/{testScope}/PortfolioManager", new PropertyValue("F Bar")),
+                            new PerpetualProperty($"Order/{testScope}/Account", new PropertyValue("J Wilson")),
+                            new PerpetualProperty($"Order/{testScope}/Strategy", new PropertyValue("RiskArb")),
```
[the remark this answers is not in the document]
Comment (timbarrass) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: It's a list of PerpetualProperties explicitly so I believe yes to metrics but no to time variant (other than the fact that you can update an order with a new set of properties).
a) I should add a test in which we retrieve and order, modify it and upsert.
b) show use of metric_values.
At this level these properties also don't need to be predefined; we should work out whether each property needs to match a property definition.

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+                        },
+                        lusidInstrument: "CCY_ZZZ"
```
[the remark this answers is not in the document]
Comment (timbarrass) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: I have another jira with suggested immediate changes to the interface. This,s trctly speaking, is not required: the client should define a st of instrument identifiers that are then resolved to an internal instrument on upsert. I'll comment for now; it shouldn't have to be here.

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+                        lusidInstrument: "CCY_ZZZ" 
+                    )
+                });
+
+            // We can ask the Orders API to upsert this order for us
+            var upsertResult = _ordersApi.UpsertOrders(scope: testScope, request: request);
+
+            // The return gives us a list of orders upserted, and the lusidinstrument for each has been mapped from the
+            // instrument identifiers passed
+            Assert.That(upsertResult.Values.Count(rl => rl.Id.Id.Equals(orderId)), Is.EqualTo(1));
```
[the remark this answers is not in the document]
Comment (timbarrass) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: I can split this into two asserts, one that the count is 1 and one that the id is as expected.

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs (hunk 1 above)
Comment (timbarrass) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: Can do.

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs (hunk 2 above)
Comment (timbarrass) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: Can do.

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs (hunk 3 above)
Comment (timbarrass) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: Can do.

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs (hunk 4 above)
Comment (timbarrass) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: Not quite sure what you mean - they're resolved to a Lusid internal instrument on upsert, so they have to exist as instruments in LUSID (or we'll resolve to unknown)?

Comment (rizwansaeed) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: Sounds like its the same. With transactions you can specify a map of identifierType:indentifierValue where the identifierValue can be of a none unique type and so long as it resolves to a single LUID all is well otherwise it resolves to ZZZ

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs (hunk 5 above)
Comment (timbarrass) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: Same reply as to Mike, above.

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs (hunk 7 above)
Comment (timbarrass) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: It returns a list of orders, each of ehich will contain the instrument identifiers passed in (note, not the full possible resolved set) and the newly resolved LusidInstrumentId

Comment (rizwansaeed) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: perhaps update the comment from `lusidinstrument` to `LusidInstrumentId` or `LUID` so its clear

Comment (timbarrass) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: Done

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (MikeMcGarry) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: Surely, the property definition needs to exist to set a value on it?

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (MikeMcGarry) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: Okay, yes it is confusing as it conflicts with the identifiers that you provide.

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (timbarrass) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: Technically no: I explicitly have to put some controls in to ensure that the property definition exists. Under the covers this isn't using the property warehouse (yet) so I suspect the definitions aren't yet critical.

For clarity/consistency though we could a) create property definitions anyway in these tutorials, then b) add the internal definitions controls in a separate jira.

[the remark this answers is not in the document]
Comment (timbarrass) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: As we move to the property warehouse .. or more likely the alternative that Guy's creating .. the definitions will be more important _I think_.

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (timbarrass) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/Orders.cs]: Fix JIRA is https://finbourne.atlassian.net/browse/DEV-6760, aiming to pick up in next sprint
